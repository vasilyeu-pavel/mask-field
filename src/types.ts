import { ChangeEvent, InputHTMLAttributes, MutableRefObject, ReactNode } from "react"

export const typeCasting = <T>(data: unknown) => data as T

interface IInput extends Omit<InputHTMLAttributes<HTMLInputElement>, "className"> {
    errors?: string[] | null;
    modifiers?: string;
}

export type TSetCursorPosition = (value: number) => void
export type TSeparator = string
export type TSeparators = TSeparator[]

export interface IChildrenProps {
    maskInputStyle: string;
    maskInputWrapperStyle: string;
    handleMaskChange: (e: ChangeEvent<HTMLInputElement>) => void;
    inputRef: MutableRefObject<HTMLInputElement>;
    value: string;
}

export type TValidator = (value: string) => string | undefined

export interface IMaskWrapperProps extends Omit<IInput, "onChange" | "value"> {
    mask: string;
    value: string;
    children: (props: IChildrenProps) => ReactNode;
    separators: TSeparators;
    onChange: (maskedValue: string, originalValue: string) => void;
    validators?: TValidator[];
    handleErrors?: (errors?: string[]) => void;
}

export interface IMaskInputProps extends
    Omit<IMaskWrapperProps, "children">
{
    type?: string;
    withErrors?: boolean;
    modifiersErrors?: string;
}
