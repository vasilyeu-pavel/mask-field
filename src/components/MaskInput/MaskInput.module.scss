.inputWrapper {
  height: 100%;
}

.errorContainer {
  display: flex;
  flex-direction: column;
  margin-top: 5px;
  font-size: 10px;
  color: #e82f3f;
}
