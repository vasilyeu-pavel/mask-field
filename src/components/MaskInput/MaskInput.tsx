import React, { memo, useState } from "react"
import { IMaskInputProps } from "../../types"
import { joinClasses } from "../../utils"
import MaskWrapper from "../MaskWrapper"

const styles = {
  inputWrapper: {
    height: "100%",
  },
  errorContainer: {
    display: "flex",
    FlexDirection: "column",
    marginTop: "5px",
    fontSize: "10px",
    color: "#e82f3f",
  },
}

export const MaskInput = ({
  type = "tel",
  value,
  separators,
  mask,
  placeholder,
  readOnly,
  disabled,
  errors,
  onChange,
  modifiers,
  modifiersErrors,
  withErrors,
  validators,
  ...props
}: IMaskInputProps) => {
  const [maskErrors, setErrors] = useState(errors)

  return (
    <div style={styles.inputWrapper}>
      <MaskWrapper
        value={value}
        separators={separators}
        mask={mask}
        errors={maskErrors}
        readOnly={readOnly}
        disabled={disabled}
        placeholder={placeholder}
        handleErrors={setErrors}
        onChange={onChange}
        modifiers={modifiers}
        validators={validators}
      >
        {({
          maskInputStyle,
          maskInputWrapperStyle,
          inputRef,
          value: maskedValue,
          handleMaskChange,
        }) => (
          <div className={maskInputWrapperStyle}>
            <input
              {...props}
              type={type}
              className={maskInputStyle}
              value={maskedValue}
              onChange={handleMaskChange}
              ref={inputRef}
              readOnly={readOnly}
              disabled={disabled}
            />
          </div>
        )}
      </MaskWrapper>
      {withErrors && maskErrors &&
        <div style={styles.errorContainer} className={joinClasses(modifiersErrors)}>
          {maskErrors.map((error, i) => (
            <span key={i}>
              {error}
            </span>
          ))}
        </div>
      }
    </div>
  )
}

export default memo(MaskInput)
