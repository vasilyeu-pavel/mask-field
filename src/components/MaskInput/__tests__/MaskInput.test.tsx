import React from "react"
import { shallow } from "enzyme"
import { IMaskInputProps } from "../../../types"
import { MaskInput } from "../MaskInput"

Object.defineProperty(document, "fonts", {
  value: {
    ready: Promise.resolve(),
  },
  writable: true,
})

describe("MaskInput", () => {
  const defaultProps: IMaskInputProps = {
    value: "value",
    separators: ["/"],
    mask: "value",
    placeholder: "placeholder",
    readOnly: false,
    errors: ["error"],
    onChange: jest.fn(),
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it("snapshot", () => {
    expect(shallow(<MaskInput {...defaultProps} />))
      .toMatchSnapshot()
  })

  test("should render error container", () => {
    const setErrors = jest.fn()

    jest.spyOn(React, "useState")
      .mockImplementationOnce(() => [["error"], setErrors])

    const component = shallow(<MaskInput {...defaultProps} withErrors />)

    expect(component.find(".errorContainer").length).toBe(1)
  })
})
