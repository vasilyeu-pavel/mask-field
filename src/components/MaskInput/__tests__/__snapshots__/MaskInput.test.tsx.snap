// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaskInput snapshot 1`] = `
<div
  className="inputWrapper"
>
  <Memo(MaskWrapper)
    errors={
      Array [
        "error",
      ]
    }
    handleErrors={[Function]}
    mask="value"
    onChange={[MockFunction]}
    placeholder="placeholder"
    readOnly={false}
    separators={
      Array [
        "/",
      ]
    }
    value="value"
  >
    <Component />
  </Memo(MaskWrapper)>
</div>
`;
